core/node/api_server/src/tx_sender/proxy.rs

```diff
@@ -1,42 +1,101 @@
 use std::{
-    collections::{BTreeSet, HashMap},
+    collections::{BTreeSet, HashMap, HashSet},
     future::Future,
     sync::Arc,
     time::Duration,
 };
 
 use anyhow::Context;
+use chrono::{TimeZone, Utc};
 use tokio::sync::{watch, RwLock};
 use zksync_dal::{
-    helpers::wait_for_l1_batch, transactions_dal::L2TxSubmissionResult, ConnectionPool, Core,
-    CoreDal,
+    helpers::wait_for_l1_batch, transactions_dal::L2TxSubmissionResult, Connection, ConnectionPool,
+    Core, CoreDal, DalError,
 };
 use zksync_shared_metrics::{TxStage, APP_METRICS};
-use zksync_types::{
-    api::{BlockId, Transaction, TransactionDetails, TransactionId},
-    fee::TransactionExecutionMetrics,
-    l2::L2Tx,
-    Address, Nonce, H256,
-};
+use zksync_types::{api, fee::TransactionExecutionMetrics, l2::L2Tx, Address, Nonce, H256, U256};
 use zksync_web3_decl::{
     client::{DynClient, L2},
     error::{ClientRpcContext, EnrichedClientResult, Web3Error},
-    namespaces::{EthNamespaceClient, ZksNamespaceClient},
+    namespaces::EthNamespaceClient,
 };
 
 use super::{tx_sink::TxSink, SubmitTxError};
 
+/// In-memory transaction cache for a full node. Works like an ad-hoc mempool replacement, with the important limitation that
+/// it's not synchronized across the network.
+///
+/// # Managing cache growth
+///
+/// To keep cache at reasonable size, the following garbage collection procedures are implemented:
+///
+/// - [`Self::run_updates()`] periodically gets nonces for all distinct accounts for the transactions in cache and removes
+///   all transactions with stale nonces. This includes both transactions included into L2 blocks and replaced transactions.
+/// - The same nonce filtering logic is applied for the transaction initiator address each time a transaction is fetched from cache.
+///   We don't want to return such transactions if they are already included in an L2 block or replaced locally, but `Self::run_updates()`
+///   hasn't run yet.
 #[derive(Debug, Clone, Default)]
 pub(crate) struct TxCache {
     inner: Arc<RwLock<TxCacheInner>>,
 }
 
 #[derive(Debug, Default)]
 struct TxCacheInner {
+    // FIXME: implement expiration / LRU?
```

Comment:
Should some kind of expiration and/or LRU logic be implemented right now? What would be an acceptable approach?